consensus/src/consensus/inner.rs

```diff
@@ -0,0 +1,508 @@
+                    }
+                    ForkDescription::Orphan => {
+                        debug!("Processing a block that is on unknown orphan chain");
+                        // dont process
+                        return Ok(());
+                    }
+                }
+            }
+        }
+        self.process_block(&hash, block).await?;
+        Ok(())
+    }
+
+    async fn commit_block(&mut self, hash: &Digest, block: &SerialBlock) -> Result<(), ConsensusError> {
+        let mut commitments = vec![];
+        let mut serials = vec![];
+        let mut memos = vec![];
+        for transaction in block.transactions.iter() {
+            commitments.extend(transaction.new_records.iter().map(|x| x.commitment.clone()));
+            serials.extend(transaction.old_serial_numbers.iter().cloned());
+            memos.push(transaction.memorandum.clone());
+        }
+        let digest = self.ledger.extend(&commitments[..], &serials[..], &memos[..])?;
+        self.storage.commit_block(hash, digest).await?;
+        Ok(())
+    }
+
+    /// removes a block and all of it's descendents from storage and ledger
+    async fn decommit_ledger_block(&mut self, hash: &Digest) -> Result<(), ConsensusError> {
+        let decommited_blocks = self.storage.decommit_blocks(hash).await?;
+        let mut commitments = vec![];
+        let mut serials = vec![];
+        let mut memos = vec![];
+        for block in decommited_blocks.into_iter() {
+            for transaction in block.transactions.iter() {
+                commitments.extend(transaction.new_records.iter().map(|x| x.commitment.clone()));
+                serials.extend(transaction.old_serial_numbers.iter().cloned());
+                memos.push(transaction.memorandum.clone());
+            }
+        }
+        self.ledger.rollback(&commitments[..], &serials[..], &memos[..])?;
+        Ok(())
+    }
+
+    /// Return whether or not the given block is valid and insert it.
+    /// 1. Verify that the block header is valid.
+    /// 2. Verify that the transactions are valid.
+    /// 3. Insert/canonize block.
+    async fn process_block(&mut self, hash: &Digest, block: &SerialBlock) -> Result<(), ConsensusError> {
+        match self.storage.get_block_state(hash).await? {
+            // todo: remove this? (extraenous)
+            BlockStatus::Committed(_) => return Ok(()),
+            BlockStatus::Unknown => return Err(ConsensusError::InvalidBlock(hash.0.to_vec())),
+            BlockStatus::Uncommitted => (),
+        }
+
+        // 1. Verify that the block valid
+        if !self.verify_block(block).await? {
+            return Err(ConsensusError::InvalidBlock(hash.0.to_vec()));
+        }
+
+        // 2. Insert/canonize block
+        self.commit_block(hash, block).await?;
+
+        // 3. Remove transactions from the mempool
+        for transaction in block.transactions.iter() {
+            self.memory_pool.remove(&transaction.id.into())?;
+        }
+
+        Ok(())
```

Comment:
To avoid confusion, I'd move the body of this method into the commit path of `try_commit_block` and remove this method altogether.

Reply:
gets called in a few places. did rename it though